twarc/decorators2.py

```diff
@@ -197,6 +197,82 @@ class InvalidAuthType(Exception):
     """
 
 
+class FileLineProgressBar(tqdm):
+    """
+    A progress bar based on input file line count. Counts an input file by lines.
+    This tries to read the entire file and count newlines in a robust way.
+    """
+
+    def __init__(self, infile, outfile, **kwargs):
+        disable = False if "disable" not in kwargs else kwargs["disable"]
+        if infile is not None and (infile.name == "<stdin>"):
+            disable = True
+        if outfile is not None and (outfile.name == "<stdout>"):
+            disable = True
+        kwargs["disable"] = disable
+        kwargs["miniters"] = 1
+        kwargs[
+            "bar_format"
+        ] = "{l_bar}{bar}| Processed {n_fmt}/{total_fmt} lines of input file [{elapsed}<{remaining}, {rate_fmt}{postfix}]"
+
+        # Warn for large (> 100 MB) input files:
+        if (os.stat(infile.name).st_size / (1024 * 1024)) > 100:
+            click.echo(
+                click.style(
+                    f"Warning: Input File Size is {os.stat(infile.name).st_size / (1024*1024):.2f} MB, it may take a while to process. CTRL+C to stop.",
+                    fg="yellow",
+                    bold=True,
+                ),
+                err=True,
+            )
+
+        def blocks(files, size=65536):
+            while True:
+                b = files.read(size)
+                if not b:
+                    break
+                yield b
+
+        total_lines = 0
+        with open(infile.name, "r", encoding="utf-8", errors="ignore") as f:
+            total_lines = sum(bl.count("\n") for bl in blocks(f))
```

Comment:
Suggestion 229 - 238 can be replaced with the following, as Python file IO in this case is on demand:

```python
with open(infile.name, "r", encoding="utf-8", errors="ignore") as f:
            total_lines = sum(1 for _ in f)
```